Add Check Support for Java 21 Pattern Matching for Switch Syntax: ParenPad

child of #14961 
I have read check documentation: https://checkstyle.org/checks/whitespace/parenpad.html#ParenPad
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
➜  src cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
        <module name="ParenPad"/>
    </module>
</module>
➜  src cat Test.java 
public class Test {
    public record Point(int x, int y) { }

    public void test(Point p) {
        switch (p) {
            case Point(int x, int y) when (  x != 0) -> System.out.println("kk"); // violation
            default -> System.out.println("k");
        }
        int x = 0;
        if (   x != 0) {};  // violation
    }
}
➜  src javac --enable-preview --release 21 Test.java
➜  src java -jar checkstyle-10.18.1-SNAPSHOT-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:6:43: '(' is followed by whitespace. [ParenPad]
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:10:12: '(' is followed by whitespace. [ParenPad]
Audit done.
Checkstyle ends with 2 errors.

```
**Describe what you expect in detail.**

This check works as expected now that we have added the `EXPR` token to the `when` expression.

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/ParenPadCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/ParenPadCheckTest.java
@@ -567,4 +567,17 @@ public class ParenPadCheckTest
         }
     }
 
+    @Test
+    public void testParenPadWithWhenExpression() throws Exception {
+        final String[] expected = {
+            "21:38: " + getCheckMessage(MSG_WS_FOLLOWED, "("),
+            "25:33: " + getCheckMessage(MSG_WS_FOLLOWED, "("),
+            "27:41: " + getCheckMessage(MSG_WS_PRECEDED, ")"),
+            "29:43: " + getCheckMessage(MSG_WS_FOLLOWED, "("),
+            "29:52: " + getCheckMessage(MSG_WS_PRECEDED, ")"),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputParenPadCheckWhenExpression.java"), expected);
+    }
+
 }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/parenpad/InputParenPadCheckWhenExpression.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/whitespace/parenpad/InputParenPadCheckWhenExpression.java
@@ -1,0 +1,38 @@
+/*
+ParenPad
+option = (default)nospace
+tokens = (default)ANNOTATION, ANNOTATION_FIELD_DEF, CTOR_CALL, CTOR_DEF, DOT, \
+         ENUM_CONSTANT_DEF, EXPR, LITERAL_CATCH, LITERAL_DO, LITERAL_FOR, LITERAL_IF, \
+         LITERAL_NEW, LITERAL_SWITCH, LITERAL_SYNCHRONIZED, LITERAL_WHILE, METHOD_CALL, \
+         METHOD_DEF, QUESTION, RESOURCE_SPECIFICATION, SUPER_CTOR_CALL, LAMBDA, RECORD_DEF
+
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.whitespace.parenpad;
+
+import java.util.HashMap;
+
+import org.w3c.dom.Node;
+
+public class InputParenPadCheckWhenExpression {
+    void test(Object o) {
+        if (o instanceof String a && ( a.length() == 9)) { }
+        // violation above, ''(' is followed by whitespace'
+
+        switch (o) {
+            case Integer i when ( i == 2) -> {}
+            // violation above, ''(' is followed by whitespace'
+            case Integer i when (i == 2 ) -> {}
+            // violation above, '')' is preceded with whitespace'
+            case Point(int x, int y) when ( x == 1 ) -> {}
+            // 2 violations above:
+            //                   ''(' is followed by whitespace'
+            //                   '')' is preceded with whitespace'
+            default -> {}
+        }
+    }
+
+    record Point(int x, int y) { }
+}

